### PR TITLE
Fix autograd engine checks and update InputMetadata

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65235

1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

Differential Revision: [D31117318](https://our.internmc.facebook.com/intern/diff/D31117318)

### CODE REVIEW

File: torch/csrc/autograd/input_metadata.h [hunk 1]
```diff
@@ -11,22 +11,29 @@
 namespace torch { namespace autograd {
 
 /**
- * Records type, shape, and device of tensor and, where applicable,
- * the stream the correspondingoperation took place on.
+ * Records TensorOptions, and shape of the tensor and, where applicable,
```
Comment (albanD) [inline torch/csrc/autograd/input_metadata.h:14]: also is python dispatch no?

File: torch/csrc/autograd/input_metadata.h [hunk 2]
```diff
@@ -11,22 +11,29 @@
...
 
-  InputMetadata(const at::TensorOptions options, at::IntArrayRef shape, at::Device device)
-  : options_{options}, shape_{shape}, device_{device} {
+  InputMetadata(const at::TensorOptions options, at::IntArrayRef shape, bool is_tensor_subclass)
+  : options_{options}, shape_{shape}, is_tensor_subclass_{is_tensor_subclass} {
+    auto device_ = options.device();
+    stream_ = c10::impl::getDeviceGuardImpl(device_.type())->getStream(device_);
+  }
+
+  InputMetadata(const at::TensorOptions options, at::IntArrayRef shape)
```
Comment (albanD) [inline torch/csrc/autograd/input_metadata.h:29]: When is this constructor called?
How is the python dispatch initialized?

File: test/test_python_dispatch.py [hunk 3]
```diff
@@ -449,5 +449,12 @@ def test_nested_enable_python_mode(self) -> None:
                 with enable_python_mode(LoggingTensor):
                     pass
 
+    def test_enable_python_mode_subclass_autograd_device_check(self) -> None:
+        x = LoggingTensor(torch.tensor([3.0, 4.0], requires_grad=True))
+        y = torch.randn(2)
+        z = x * y
+        assert isinstance(z, LoggingTensor)
```
Comment (albanD) [inline test/test_python_dispatch.py:456]: self.assertIsInstance

File: test/test_python_dispatch.py [hunk 4]
```diff
@@ -449,5 +449,12 @@ def test_nested_enable_python_mode(self) -> None:
...
+        z.sum().backward(torch.tensor(1.))
```
Comment (albanD) [inline test/test_python_dispatch.py:457]: Can you check that the computed gradient is correct?

**Comment by ezyang** [comment]
I'm letting Alban do full review on this.

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -701,22 +695,31 @@ void validate_outputs(
...
+       ss << "invalid gradient at index " << i << " - expected dtype ";
+       ss << metadata.dtype() << " but got " << grad.dtype();
+       AT_ERROR(format_error(ss.str()));
     }
-    if (!is_compatible_type(metadata.options(), grad.options())) {
+    if (grad.layout() != metadata.layout()) {
        std::stringstream ss;
-       ss << "invalid gradient at index " << i << " - expected type ";
-       ss << metadata.options() << " but got " << grad.options();
+       ss << "invalid gradient at index " << i << " - expected layout ";
```
Comment (anjali411) [inline torch/csrc/autograd/engine.cpp:706]: There are currently four layouts: sparse, sparsecsr, mkldnn and strided. Currently, the only cross combination of layouts for tensor and grad that's supported is `(Strided, Sparse)` (based on the check that was being done by `is_compatible_type`). I am not sure about mkldn but it seems like we should support all possible combinations of (tensor.layout(), tensor.grad.layout()) for the other three layouts. cc. @albanD

Comment (albanD) [reply torch/csrc/autograd/engine.cpp:706]: I think we can keep this current layout logic as-is and only update the subclass case.

File: torch/csrc/autograd/input_metadata.h (hunk 2 above)
Comment (anjali411) [reply torch/csrc/autograd/input_metadata.h]: removed it

**Review by albanD — COMMENTED** [review]
sgtm just the continue thing

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by anjali411** [comment]
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31082693).

**Comment by anjali411** [comment]
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31117318).

**Comment by anjali411** [comment]
@ezyang @albanD should we cherry-pick this in 1.10?

**Comment by albanD** [comment]
I would say no since I don't think we plan to have broad subclassing support in 1.10 anyways?
